src/pal/src/thread/thread.cpp

```diff
@@ -1578,7 +1578,124 @@ GetThreadTimes(
     return (retval);
 }
 
+HRESULT
+PALAPI
+SetThreadDescription(    
+    IN HANDLE hThread,
+    IN PCWSTR lpThreadDescription)
+{
+    CPalThread *pThread;
+    PAL_ERROR palError;
+
+    PERF_ENTRY(SetThreadDescription);
+    ENTRY("SetThreadDescription(hThread=%p,lpThreadDescription=%p)\n", hThread, lpThreadDescription);
+
+    pThread = InternalGetCurrentThread();
+
+    palError = InternalSetThreadDescription(
+        pThread,
+        hThread,
+        lpThreadDescription
+        );
+
+    if (NO_ERROR != palError)
+    {
+        pThread->SetLastError(palError);
+    }
+
+    LOGEXIT("SetThreadDescription");
+    PERF_EXIT(SetThreadDescription);
+
+    return HRESULT_FROM_WIN32(palError);
+}
 
+PAL_ERROR
+CorUnix::InternalSetThreadDescription(
+    CPalThread *pThread,
+    HANDLE hTargetThread,
+    PCWSTR lpThreadDescription
+)
+{
+    PAL_ERROR palError = NO_ERROR;
+    CPalThread *pTargetThread = NULL;
+    IPalObject *pobjThread = NULL;
+    int error;
+    int nameSize;
+    char *nameBuf = NULL;
+
+// The exact API of pthread_setname_np varies very wildly depending on OS.
+// For now, only Linux is implemented.
+#if defined(__linux__)
+
+    palError = InternalGetThreadDataFromHandle(
+        pThread,
+        hTargetThread,
+        &pTargetThread,
+        &pobjThread
+        );
+
+    if (NO_ERROR != palError)
+    {
+        goto InternalSetThreadDescriptionExit;
+    }
+
+    pTargetThread->Lock(pThread);
+
+    /* translate the wide char lpThreadDescription string to multibyte string */
+    nameSize = WideCharToMultiByte(CP_ACP, 0, lpThreadDescription, -1, NULL, 0, NULL, NULL);
+
+    if (0 == nameSize)
+    {
+        palError = ERROR_INTERNAL_ERROR;
+        goto InternalSetThreadDescriptionExit;
+    }
+
+    nameBuf = (char *)PAL_malloc(nameSize);
```

Comment:
Question: since nameBuf will be truncated to 16 bytes, can you allocate this on the stack instead?  I don't know how WideCharToMultiByte() will behave if truncation is necessary, though...

Reply:
That was my first thought too, but when I looked at `WideCharToMultiByte` it seemed like there was no way to make it truncate. Maybe I missed something in the documentation though.

Reply:
WideCharToMultiByte fails with ERROR_INSUFFICIENT_BUFFER if the target buffer is not large enough for the source string to encode.

Reply:
Yeah, maybe leave it as is.  This shouldn't be a hot function anyway.